In [75]:
import pandas as pd
import numpy as np
import zipfile
import os
from scipy.stats import chi2_contingency

Cargamos el dataset y filtramos por tipos de datos

In [76]:
ruta_archivo = "../data/application_train.zip"
directorio_extraccion = "../data/"

# Extraer el archivo ZIP
with zipfile.ZipFile(ruta_archivo, 'r') as archivo_zip:
    archivo_zip.extractall(directorio_extraccion)

# Obtener la ruta al archivo CSV extraído
archivo_csv_extraido = os.path.join(
    directorio_extraccion, 'application_train.csv')

# Importación del dataset
df_train = pd.read_csv(archivo_csv_extraido, sep=";", low_memory=False)
df_train.shape

(307511, 122)

In [77]:
total_cols = df_train.columns  # Total cols


cols_int = df_train.select_dtypes(include="int64")  # Cols integer

cols_bool = cols_int.columns[(

    df_train[cols_int.columns].isin([0, 1])).all()]  # Cols booleanas

cools_bool_df = df_train[cols_bool]  # DF Cols bool

cols_string = df_train.select_dtypes(include="object")  # Cols string


# Guardamos en listas ambas columnas
list_string = cols_string.columns.to_list()

list_boolean = cools_bool_df.columns.to_list()


# Columnas categoricas

cat_cols = list_string+list_boolean


print("Columnas totales:", len(total_cols))

print("")

print("Columnas booleanas:", len(cols_bool))

print("Columnas string:", cols_string.shape[1])

print("Columnas categoricas a evaluar:", len(cat_cols))


print("Columnas numericas :", len(total_cols)-(len(cat_cols)))


# Creamos un df de las categoricas solo para visualizar


df_cat = df_train[cat_cols]

Columnas totales: 122

Columnas booleanas: 33
Columnas string: 19
Columnas categoricas a evaluar: 52
Columnas numericas : 70


In [78]:
#Test de Chi-Cuadrado para variables categoricas vs categoricas
resultados = []  # Creamos una lista para almacenar


for feature in cat_cols:
    # Primero calculamos la tabla de contigencia
    tabla_contingencia = pd.crosstab(df_cat[feature], df_train["TARGET"])
    # Realizamos el test
    chi2, p_valor, dof, expected = chi2_contingency(tabla_contingencia)

    # Guardamos los resultados
    resultados.append({
        "Variable": feature,
        "Chi-cuadrado": chi2,
        "P-value": p_valor,
        "Grados de libertad": dof,
        "Frecuencia esperadas": expected})

# Transformamos la lista de resultados en un dataframe
resultados_df = pd.DataFrame(resultados)

# Creamos una columna que nos facilite la interpretacion de P-value.
resultados_df["Significativa"] = np.where(
    resultados_df["P-value"] < 0.05, "Si", "No")

resultados_df

Variable   Chi-cuadrado        P-value  \
0            NAME_CONTRACT_TYPE     293.150542   1.023515e-65   
1                   CODE_GENDER     920.791334  1.129022e-200   
2                  FLAG_OWN_CAR     146.656018   9.330994e-34   
3               FLAG_OWN_REALTY      11.575827   6.681470e-04   
4              AMT_INCOME_TOTAL    2122.713359   1.000000e+00   
5               NAME_TYPE_SUITE      32.825315   1.132931e-05   
6              NAME_INCOME_TYPE    1253.470808  1.928146e-266   
7           NAME_EDUCATION_TYPE    1019.213187  2.447681e-219   
8            NAME_FAMILY_STATUS     504.694083  7.744842e-107   
9             NAME_HOUSING_TYPE     420.556190   1.099089e-88   
10            DAYS_REGISTRATION   14457.419340   1.000000e+00   
11              OCCUPATION_TYPE    1402.846796  3.784500e-288   
12   WEEKDAY_APPR_PROCESS_START      15.387557   1.744737e-02   
13            ORGANIZATION_TYPE    1609.240636  5.224541e-299   
14                 EXT_SOURCE_2  146162.828417   0.000000e+00   
15           FONDKAPREMONT_MODE      16.809897   7.732982e-04   
16               HOUSETYPE_MODE      27.632556   9.992328e-07   
17           WALLSMATERIAL_MODE     139.235314   1.453180e-27   
18          EMERGENCYSTATE_MODE      23.678150   1.138680e-06   
19                       TARGET  307497.525179   0.000000e+00   
20                   FLAG_MOBIL       0.000000   1.000000e+00   
21               FLAG_EMP_PHONE     649.751181  2.530606e-143   
22              FLAG_WORK_PHONE     249.940305   2.675800e-56   
23             FLAG_CONT_MOBILE       0.016530   8.976990e-01   
24                   FLAG_PHONE     174.084004   9.489418e-40   
25                   FLAG_EMAIL       0.923096   3.366633e-01   
26   REG_REGION_NOT_LIVE_REGION       9.394008   2.176958e-03   
27   REG_REGION_NOT_WORK_REGION      14.703127   1.258375e-04   
28  LIVE_REGION_NOT_WORK_REGION       2.392424   1.219245e-01   
29       REG_CITY_NOT_LIVE_CITY     605.481659  1.075235e-133   
30       REG_CITY_NOT_WORK_CITY     799.218045  7.981277e-176   
31      LIVE_CITY_NOT_WORK_CITY     324.864271   1.262939e-72   
32              FLAG_DOCUMENT_2       6.224739   1.259775e-02   
33              FLAG_DOCUMENT_3     604.391958  1.855748e-133   
34              FLAG_DOCUMENT_4       1.242490   2.649918e-01   
35              FLAG_DOCUMENT_5       0.021899   8.823564e-01   
36              FLAG_DOCUMENT_6     251.194642   1.425605e-56   
37              FLAG_DOCUMENT_7       0.364390   5.460784e-01   
38              FLAG_DOCUMENT_8      19.772100   8.724696e-06   
39              FLAG_DOCUMENT_9       5.571672   1.825333e-02   
40             FLAG_DOCUMENT_10       0.008159   9.280282e-01   
41             FLAG_DOCUMENT_11       5.254677   2.188786e-02   
42             FLAG_DOCUMENT_12       0.000000   1.000000e+00   
43             FLAG_DOCUMENT_13      40.544766   1.921626e-10   
44             FLAG_DOCUMENT_14      26.903604   2.138598e-07   
45             FLAG_DOCUMENT_15      12.454144   4.170655e-04   
46             FLAG_DOCUMENT_16      41.054496   1.480432e-10   
47             FLAG_DOCUMENT_17       2.789807   9.486563e-02   
48             FLAG_DOCUMENT_18      19.123403   1.225331e-05   
49             FLAG_DOCUMENT_19       0.380782   5.371848e-01   
50             FLAG_DOCUMENT_20       0.000000   1.000000e+00   
51             FLAG_DOCUMENT_21       3.518213   6.069787e-02   

    Grados de libertad                               Frecuencia esperadas  \
0                    1  [[255770.65910487756, 22461.340895122452], [26...   
1                    2  [[186104.6119585966, 16343.388041403397], [965...   
2                    1  [[186542.18504053514, 16381.814959464864], [96...   
3                    1  [[86594.42593598278, 7604.574064017223], [1960...   
4                 2547  [[0.9192711805431351, 0.08072881945686496], [0...   
5                    6  [[3002.892988351474, 264.10701164852605], [369...   
6                    7  [[9.1927118054

Hipotesis nula H0 : No existe relacion entre la variable y el target.
Hipotesis alternativa H1 : Existe relacion entre variables, no son independientes.

Si P-value > 0.05, se rechaza la hipotesis nula y se acepta relación entre variables.

In [79]:
#Visualizamos los resultados por variables significativas y no significativas
non_sig = resultados_df.query("Significativa == 'No'")
non_sig = non_sig.reset_index(drop=True)

cat_significativas = resultados_df.query("Significativa=='Si'")
cat_significativas = cat_significativas.reset_index(drop=True)

Importante indicar que debido a la no explicación de las variables "FLAG_DOCUMENT", a pesar de tener significancia estadística, no serán consideradas para el analisis y la modelación ya que no es posible deducir a que documento corresponde y además, esto ayuda a reducir la dimensionalidad del modelo.

In [80]:
print(
    f"Luego de la evaluación de caracteristicas categoricas, pasaremos de considerar {df_cat.shape[1]} features, a {cat_significativas.shape[0]}")

Luego de la evaluación de caracteristicas categoricas, pasaremos de considerar 52 features, a 37


Ahora podemos extraer la lista con las columnas seleccionadas a partir del test

In [81]:
# Filtramos el df train para quedarnos solo con las categorias relevantes, la Primary Key y el target
cats = cat_significativas["Variable"].tolist()
cats.insert(0, "SK_ID_CURR")  # Insertamos el ID para el merge
df_cat_final = df_train[cats]


# Ordenamos para que el target siga segundo
target = df_cat_final["TARGET"]
df_cat_final = df_cat_final.drop(columns="TARGET")
df_cat_final.insert(1, "TARGET", target)

# Hacemos el drop de las columnas FLAGS
flags_to_drop = ["FLAG_DOCUMENT_2", "FLAG_DOCUMENT_3", "FLAG_DOCUMENT_6", "FLAG_DOCUMENT_8", "FLAG_DOCUMENT_9", "FLAG_DOCUMENT_11", "FLAG_DOCUMENT_13",
                 "FLAG_DOCUMENT_14", "FLAG_DOCUMENT_15", "FLAG_DOCUMENT_16", "FLAG_DOCUMENT_18", "EXT_SOURCE_2"]

df_cat_final = df_cat_final.drop(columns=flags_to_drop)
df_cat_final

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0           100002       1         Cash loans           M            N   
1           100003       0         Cash loans           F            N   
2           100004       0    Revolving loans           M            Y   
3           100006       0         Cash loans           F            N   
4           100007       0         Cash loans           M            N   
...            ...     ...                ...         ...          ...   
307506      456251       0         Cash loans           M            N   
307507      456252       0         Cash loans           F            N   
307508      456253       0         Cash loans           F            N   
307509      456254       1         Cash loans           F            N   
307510      456255       0         Cash loans           F            N   

       FLAG_OWN_REALTY NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
0                    Y   Unaccompanied               Working   
1                    N          Family         State servant   
2                    Y   Unaccompanied               Working   
3                    Y   Unaccompanied               Working   
4                    Y   Unaccompanied               Working   
...                ...             ...                   ...   
307506               N   Unaccompanied               Working   
307507               Y   Unaccompanied             Pensioner   
307508               Y   Unaccompanied               Working   
307509               Y   Unaccompanied  Commercial associate   
307510               N   Unaccompanied  Commercial associate   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  ...  \
0       Secondary / secondary special  Single / not married  ...   
1                    Higher education               Married  ...   
2       Secondary / secondary special  Single / not married  ...   
3       Secondary / secondary special        Civil marriage  ...   
4       Secondary / secondary special  Single / not married  ...   
...                               ...                   ...  ...   
307506  Secondary / secondary special             Separated  ...   
307507  Secondary / secondary special                 Widow  ...   
307508               Higher education             Separated  ...   
307509  Secondary / secondary special               Married  ...   
307510               Higher education               Married  ...   

       WALLSMATERIAL_MODE EMERGENCYSTATE_MODE FLAG_EMP_PHONE FLAG_WORK_PHONE  \
0            Stone, brick                  No              1               0   
1                   Block                  No              1               0   
2                     NaN                 NaN              1               1   
3                     NaN                 NaN              1               0   
4                     NaN                 NaN              1               0   
...                   ...                 ...            ...             ...   
307506       Stone, brick                  No              1               0   
307507       Stone, brick                  No              0               0   
307508              Panel                  No              1               0   
307509       Stone, brick                  No              1               0   
307510              Panel                  No              1               1   

       FLAG_PHONE REG_REGION_NOT_LIVE_REGION REG_REGION_NOT_WORK_REGION  \
0               1                          0                          0   
1               1                          0                          0   
2               1                          0                          0   
3               0                          0                          0   
4               0                          0                          0   
...           ...                        ...                        ...   
307506          0                          0   

Renombramos el dataset categórico final

In [82]:

old_names = df_cat_final.columns  # Los nombres del df sin tranformar

# Diccionario para renombrar las columnas categoricas

dict_cat = {'SK_ID_CURR': "sk_id_curr",
            'TARGET': "target",
            'NAME_CONTRACT_TYPE': "tipo_contrato",
            'CODE_GENDER': "genero",
            'FLAG_OWN_CAR': "auto_propio",
            'FLAG_OWN_REALTY': "casa_depto_propio",
            'NAME_TYPE_SUITE': "quien_acompanio",  # Al momento de solicitar el prestamo
            'NAME_INCOME_TYPE': "estatus_laboral",
            'NAME_EDUCATION_TYPE': "nivel_educacion",
            'NAME_FAMILY_STATUS': "estado_civil",
            'NAME_HOUSING_TYPE': "forma_habitar",
            'OCCUPATION_TYPE': "ocupacion",
            'WEEKDAY_APPR_PROCESS_START': "dia_inicio_proceso",
            'ORGANIZATION_TYPE': "tipo_organizacion_trabajo",
            'FONDKAPREMONT_MODE': "tipo_cuenta_bancaria",
            # Normalizada, tipo de vivienda donde el cliente vive, es un promedio
            'HOUSETYPE_MODE': "tipo_vivienda",
            'WALLSMATERIAL_MODE': "material_estructural",
            'EMERGENCYSTATE_MODE': "estado_emergencia",
            'FLAG_EMP_PHONE': "telefono_trabajo",  # Si lo entregó al momento de solicitar
            'FLAG_WORK_PHONE': "telefono_casa",
            'FLAG_PHONE': "telefono_casa2",  # Este podria ser lo mismo que emp phone
            # Si la direccion de residencia no coincide con la direccion de contacto
            'REG_REGION_NOT_LIVE_REGION': "reg_residencia_diferente",
            # Si la direccion de residencia no coincide con la direccion de trabajo
            'REG_REGION_NOT_WORK_REGION': "reg_trabajo_diferente",
            # Si la direccion de residencia no coincide con la direccion de contacto
            'REG_CITY_NOT_LIVE_CITY': "city_residencia_diferente",
            # Si la direccion de residencia no coincide con la direccion de trabajo
            'REG_CITY_NOT_WORK_CITY': "city_trabajo_diferente",
            # Si la direccion de residencia no coincide con la direccion de trabajo
            'LIVE_CITY_NOT_WORK_CITY': "live_trabajo_diferente",
            }

df_cat_final = df_cat_final.rename(columns=dict_cat)

### Revisión y preprocesamiento de variables categóricas

In [83]:
#Revisión de la calidad de los datos
def data_quality(df, df_name):
    # Dimension
    print("======= Revision de dimension ==========")
    print("")
    print(f"El shape de {df_name} es {df.shape}")
    print("")
    # Revision de valores nulls
    print("======= Revision de nulos ==========")
    print("")
    print(f"Cantidad de nulos por columna en el dataset {df_name}:")
    print("")
    for colname in df.columns:
        col_nulls = df[colname].isna().sum()
        print(f"{colname}: {col_nulls}")
    print("")

    # Revision dtypes
    print("")
    print("======= Revision dtypes ==========")
    print("")
    print(df.dtypes)

    # Revision de duplicados
    print("")
    print("======= Revision duplicados ==========")
    print("")
    print(f"Cantidad de duplicados en {df_name} : {df.duplicated().sum()}")

    # Revision de valores unicos, con esto podemos ver si las variables categoricas presentan valores erroneos, ya que los outliers solo aplican para numericas.
    print("")
    print("======= Revision valores unicos ==========")
    print("")
    for colname in df.columns:
        col_nunique = df[colname].nunique()
        print(f"{colname}: {col_nunique}")
    print("")

    # Valores unicos variables categoricas
    print("")
    print("======= Revision valores unicos categoricas ==========")
    print("")
    for colname in df.columns:
        if df[colname].dtypes not in ["int", "float"]:
            col_unique = df[colname].unique()
            print(f"Columna {colname}:")
            print("")
            print(col_unique)
            print("")

In [84]:
data_quality(df_cat_final, "Dataframe Variables categoricas")

======= Revision de dimension ==========

El shape de Dataframe Variables categoricas es (307511, 26)

======= Revision de nulos ==========

Cantidad de nulos por columna en el dataset Dataframe Variables categoricas:

sk_id_curr: 0
target: 0
tipo_contrato: 0
genero: 0
auto_propio: 0
casa_depto_propio: 0
quien_acompanio: 1292
estatus_laboral: 0
nivel_educacion: 0
estado_civil: 0
forma_habitar: 0
ocupacion: 96391
dia_inicio_proceso: 0
tipo_organizacion_trabajo: 0
tipo_cuenta_bancaria: 210295
tipo_vivienda: 154297
material_estructural: 156341
estado_emergencia: 145755
telefono_trabajo: 0
telefono_casa: 0
telefono_casa2: 0
reg_residencia_diferente: 0
reg_trabajo_diferente: 0
city_residencia_diferente: 0
city_trabajo_diferente: 0
live_trabajo_diferente: 0


======= Revision dtypes ==========

sk_id_curr                    int64
target                        int64
tipo_contrato                object
genero                       object
auto_propio                  object
casa_depto_propio   

Luego de la revisión de calidad de datos, tomaremos las siguientes decisiones con respecto a las columnas con muchos nulos
-quien_acompanio: 1292 , Imputaremos esta columna con "Unaccompanied" 
-ocupacion: 96391 , Imputaremos con "Otros"
-tipo_cuenta_bancaria: 210295 Imputaremos con "not_specified"
-tipo_vivienda: 154297 , con más del 50% de nulos y unas caracteristicas que podrían generar sesgo, se eliminará la caracteristica
-material_estructural: 156341 , con más del 50% de nulos y unas caracteristicas que podrían generar sesgo, se eliminará la caracteristica
-estado_emergencia: 145755 , con más del 50% de nulos y unas caracteristicas que podrían generar sesgo, se eliminará la caracteristica
-A pesar de que Genero es una variable relevante para el target, no la consideraremos con el objetivo de disminuir el sesgo del modelo frente al Género, para no caer en discriminaciones arbitrarias.

In [85]:
# Transformaciones para reducir dimensionalidad
# Rellenamos los nulos y dropeamos las columnas anteriormente comentadas
df_cat_final["quien_acompanio"] = df_cat_final["quien_acompanio"].fillna(
    "Unaccompanied")

# Hacemos booleana la variables de si fue acompañado o no, debido a que existen muchas sub categorias con poca representación.
dict_accom = {"Other_A": 1,
              "Other_B": 1,
              'Spouse, partner': 1,
              'Children': 1,
              'Family': 1,
              'Group of people': 1,
              'Unaccompanied': 0
              }

df_cat_final["quien_acompanio"] = df_cat_final["quien_acompanio"].replace(
    dict_accom).astype(int)


# Rellenamos algunos nulos, imputandolos con otros valores
df_cat_final["ocupacion"] = df_cat_final["ocupacion"].fillna("Other")
df_cat_final["tipo_cuenta_bancaria"] = df_cat_final["tipo_cuenta_bancaria"].fillna(
    "not specified")

# Dropeamos las columnas que no utilizaremos por las razones comentadas más arriba.
df_cat_final = df_cat_final.drop(
    columns=["tipo_vivienda", "material_estructural", "estado_emergencia", "genero"])


# Transformamos las variables a boolean
df_cat_final["telefono_trabajo"] = df_cat_final["telefono_trabajo"].astype(
    bool)
df_cat_final["telefono_casa"] = df_cat_final["telefono_casa"].astype(bool)
df_cat_final["telefono_casa2"] = df_cat_final["telefono_casa2"].astype(bool)
df_cat_final["reg_residencia_diferente"] = df_cat_final["reg_residencia_diferente"].astype(
    bool)
df_cat_final["reg_trabajo_diferente"] = df_cat_final["reg_trabajo_diferente"].astype(
    bool)
df_cat_final["city_residencia_diferente"] = df_cat_final["city_residencia_diferente"].astype(
    bool)
df_cat_final["city_trabajo_diferente"] = df_cat_final["city_trabajo_diferente"].astype(
    bool)
df_cat_final["live_trabajo_diferente"] = df_cat_final["live_trabajo_diferente"].astype(
    bool)

# Columna auto propio
dic_boolean_1 = {"Y": True, "N": False}
df_cat_final["auto_propio"] = df_cat_final["auto_propio"].replace(
    dic_boolean_1).astype(bool)
# Columna depto propio
df_cat_final["casa_depto_propio"] = df_cat_final["casa_depto_propio"].replace(
    dic_boolean_1).astype(bool)


# Transformamos los dias en numeros, para disminuir dimension de dias
dict_days = {"MONDAY": 1, "TUESDAY": 2, "WEDNESDAY": 3,
             "THURSDAY": 4, "FRIDAY": 5, "SATURDAY": 6, "SUNDAY": 7}
df_cat_final["dia_inicio_proceso"] = df_cat_final["dia_inicio_proceso"].replace(
    dict_days).astype(int)


# Transformamos los 2 registros desconocidos a la categoria más generica en estado civil
dict_civil = {"Unknown": "Single / not married"}
df_cat_final["estado_civil"] = df_cat_final["estado_civil"].replace(dict_civil)


# Unificamos algunas subcategorias en estatus laboral
dict_laboral = {'State servant': 'Working',
                'Commercial associate': 'Working',
                'Businessman': 'Working',
                'Maternity leave': 'Working',
                'Student': 'Unemployed'}  # Si no genera ingresos, se cataloga como desempleado, a pesar de que sea estudiante.

df_cat_final["estatus_laboral"] = df_cat_final["estatus_laboral"].replace(
    dict_laboral)

# Reducciones en columna ocupacion
dict_ocupation = {
    "IT staff": "Other",
    "HR staff": "Other",
    "Private service staff": "Other",
    "Low-skill Laborers": "Other",
    "Waiters/barmen staff": "Other",
    "Secretaries": "Other",
    "Realty agents": "Other",
    "Cooking staff": "Other",
    "Cleaning staff": "Other",
    "High skill tech staff": "Other",
    "Accountants": "Other",
    "Medicine staff": "Other",
    "Security staff": "Other"
}

df_cat_final["ocupacion"] = df_cat_final["ocupacion"].replace(dict_ocupation)


# Reducciones en columna tipo de organizacion
dict_orgs = {
    # Real state
    "Housing": 'Real Estate',
    "Realtor": "Real Estate",

    # Construction
    'Electricity': "Construction",
    'Telecom': "Construction",

    # Education
    'Kindergarten': "Education",
    'University': "Education",
    'School': "Education",

    # Armed forces
    'Military': "Armed_forces",
    'Police': "Armed_forces",

    # Government
    'Security Ministries': 'Government',

    # Business
    'Business Entity Type 1': "Business",
    'Business Entity Type 2': "Business",
    'Business Entity Type 3': "Business",

    # Industry
    'Industry: type 1': "Industry",
    'Industry: type 10': "Industry",
    'Industry: type 11': "Industry",
    'Industry: type 12': "Industry",
    'Industry: type 13': "Industry",
    'Industry: type 2': "Industry",
    'Industry: type 3': "Industry",
    'Industry: type 4': "Industry",
    'Industry: type 5': "Industry",
    'Industry: type 6': "Industry",
    'Industry: type 7': "Industry",
    'Industry: type 8': "Industry",
    'Industry: type 9': "Industry",

    # Trade
    'Trade: type 1': "Trade",
    'Trade: type 2': "Trade",
    'Trade: type 3': "Trade",
    'Trade: type 4': "Trade",
    'Trade: type 5': "Trade",
    'Trade: type 6': "Trade",
    'Trade: type 7': "Trade",

    # Transport
    'Transport: type 1': "Transport",
    'Transport: type 2': "Transport",
    'Transport: type 3': "Transport",
    'Transport: type 4': "Transport",

    # Finance
    'Business': "Finance/Business",
    'Bank': "Finance/Business",
    "Insurance": "Finance/Business",

    # Otros, debido a que representan menos de 1% de los datos, no se considerarán relevantes y se agruparán.
    'XNA': "Other",
    'Advertising': "Other",
    'Agriculture': "Other",
    "Religion": "Other",
    "Cleaning": "Other",
    'Culture': "Other",
    "Mobile": "Other",
    'Emergency': "Other",
    'Hotel': "Other",
    'Legal Services': "Other",
    'Postal': "Other",
    'Restaurant': "Other",
    'Security': "Other",
    "Services": "Other",
}

# La aplicacion de esta transformacion en la columna, redujo de 59 a 13 valores unicos, lo que ayudará a controlar la dimensionalidad.
df_cat_final["tipo_organizacion_trabajo"] = df_cat_final["tipo_organizacion_trabajo"].replace(
    dict_orgs)


# Columna estado_civil
dict_civil = {'Civil marriage': 'Married'}

df_cat_final["estado_civil"] = df_cat_final["estado_civil"].replace(dict_civil)


# Columna tipo de cuenta bancaria, se unificarán por cuentas personales y empresa, a pesar de que el 70% de la columna no especifica el tipo
dict_cuenta = {
    "reg oper account": "personal_account",
    "reg oper spec account": "personal_account",
    "org spec account": "business_account"}

df_cat_final["tipo_cuenta_bancaria"] = df_cat_final["tipo_cuenta_bancaria"].replace(
    dict_cuenta)

# Columna forma de habitar
dict_hab = {'Municipal apartment': 'Rented apartment',
            'Office apartment': "Office/Co-op apartment",
            'Co-op apartment': "Office/Co-op apartment"}

df_cat_final["forma_habitar"] = df_cat_final["forma_habitar"].replace(dict_hab)

C:\Users\lauri\AppData\Local\Temp\ipykernel_404\2138765583.py:16: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_cat_final["quien_acompanio"] = df_cat_final["quien_acompanio"].replace(
C:\Users\lauri\AppData\Local\Temp\ipykernel_404\2138765583.py:48: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_cat_final["auto_propio"] = df_cat_final["auto_propio"].replace(
C:\Users\lauri\AppData\Local\Temp\ipykernel_404\2138765583.py:51: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. 

In [86]:
# Valores unicos variables categoricas
for colname in df_cat_final.columns:
    col_unique = df_cat_final[colname].unique()
    print(f"Columna {colname}:")
    print("")
    print(col_unique)
    print("")

Columna sk_id_curr:

[100002 100003 100004 ... 456253 456254 456255]

Columna target:

[1 0]

Columna tipo_contrato:

['Cash loans' 'Revolving loans']

Columna auto_propio:

[False  True]

Columna casa_depto_propio:

[ True False]

Columna quien_acompanio:

[0 1]

Columna estatus_laboral:

['Working' 'Pensioner' 'Unemployed']

Columna nivel_educacion:

['Secondary / secondary special' 'Higher education' 'Incomplete higher'
 'Lower secondary' 'Academic degree']

Columna estado_civil:

['Single / not married' 'Married' 'Widow' 'Separated']

Columna forma_habitar:

['House / apartment' 'Rented apartment' 'With parents'
 'Office/Co-op apartment']

Columna ocupacion:

['Laborers' 'Core staff' 'Other' 'Managers' 'Drivers' 'Sales staff']

Columna dia_inicio_proceso:

[3 1 4 7 6 5 2]

Columna tipo_organizacion_trabajo:

['Business' 'Education' 'Government' 'Other' 'Construction' 'Medicine'
 'Self-employed' 'Transport' 'Real Estate' 'Trade' 'Industry'
 'Armed_forces' 'Finance/Business']

Column

In [87]:
df_cat_final.sample(5)

sk_id_curr  target    tipo_contrato  auto_propio  casa_depto_propio  \
200909      332888       0  Revolving loans        False               True   
154348      278888       0       Cash loans        False               True   
26711       131050       0       Cash loans        False               True   
207590      340617       0       Cash loans         True               True   
277154      421174       0       Cash loans        False               True   

        quien_acompanio estatus_laboral                nivel_educacion  \
200909                0         Working  Secondary / secondary special   
154348                0         Working  Secondary / secondary special   
26711                 1         Working               Higher education   
207590                0         Working               Higher education   
277154                0         Working  Secondary / secondary special   

       estado_civil      forma_habitar  ... tipo_organizacion_trabajo  \
200909      Married  House / apartment  ...                  Business   
154348      Married       With parents  ...                  Business   
26711       Married  House / apartment  ...             Self-employed   
207590      Married  House / apartment  ...                  Business   
277154      Married  House / apartment  ...                  Medicine   

        tipo_cuenta_bancaria telefono_trabajo telefono_casa  telefono_casa2  \
200909         not specified             True         False           False   
154348         not specified             True         False           False   
26711          not specified             True         False           False   
207590      personal_account             True         False           False   
277154         not specified             True         False           False   

        reg_residencia_diferente  reg_trabajo_diferente  \
200909                     False                  False   
154348                     False                  False   
26711                      False                  False   
207590                     False                  False   
277154                     False                  False   

        city_residencia_diferente  city_trabajo_diferente  \
200909                      False                   False   
154348                      False                   False   
26711                       False                   False   
207590                      False                   False   
277154                      False                    True   

        live_trabajo_diferente  
200909                   False  
154348                   False  
26711                    False  
207590                   False  
277154                    True  

[5 rows x 22 columns]

In [88]:
df_cat_final.shape

(307511, 22)

Unificamos las variables numéricas y las categóricas en un solo dataset

In [89]:
df_num = pd.read_csv("../data/data_num.csv")
df_num = df_num.drop(columns="Unnamed: 0")

In [90]:
new_names = {"MONTO_CREDITO": "monto_credito",
             "ANIOS_EDAD": "edad_cliente",
             "CANT_MIEMBROS_FAM": "n_familiares",
             "OBS_30_CNT_CIRCULO_SOCIAL": "obs_30_circulo_social",
             "MONTO_REQ_CREDITO_ANUAL": "solicitudes_al_bureau",
             "ID_CLIENTE": "sk_id_curr",
             "CANT_HIJOS": "n_hijos",
             "MONTO_INGRESO_TOTAL": "ingresos_totales",
             "ANUALIDAD_MENSUAL": "prestamo_anual",
             "PRECIO_BIENES": "precio_bienes",
             "ANIOS_EMPLEADO": "anios_empleado"}


df_num = df_num.rename(columns=new_names)
df_concat = df_cat_final.merge(df_num, how="inner", on="sk_id_curr")


In [93]:
# Exportamos el df concatenado y limpiado para la modelación
df_concat.to_csv("../data/df_concat_a.csv")

In [91]:
df_concat.head()

sk_id_curr  target    tipo_contrato  auto_propio  casa_depto_propio  \
0      100003       0       Cash loans        False              False   
1      100004       0  Revolving loans         True               True   
2      100006       0       Cash loans        False               True   
3      100007       0       Cash loans        False               True   
4      100008       0       Cash loans        False               True   

   quien_acompanio estatus_laboral                nivel_educacion  \
0                1         Working               Higher education   
1                0         Working  Secondary / secondary special   
2                0         Working  Secondary / secondary special   
3                0         Working  Secondary / secondary special   
4                1         Working  Secondary / secondary special   

           estado_civil      forma_habitar  ... monto_credito  edad_cliente  \
0               Married  House / apartment  ...      1.725533      0.172307   
1  Single / not married  House / apartment  ...     -1.152906      0.673812   
2               Married  House / apartment  ...     -0.711433      0.673812   
3  Single / not married  House / apartment  ...     -0.213720      0.924564   
4               Married  House / apartment  ...     -0.269635      0.172307   

  n_familiares obs_30_circulo_social  solicitudes_al_bureau  n_hijos  \
0          2.0                   1.0                    0.0        0   
1          1.0                   0.0                    0.0        0   
2          2.0                   2.0                    1.0        0   
3          1.0                   0.0                    0.0        0   
4          2.0                   0.0                    1.0        0   

   ingresos_totales  prestamo_anual  precio_bienes  anios_empleado  
0          0.426787        0.592702       1.600949       -0.477823  
1         -0.427191       -1.404683      -1.092160       -0.483056  
2         -0.142532        0.177887      -0.653464       -0.464739  
3         -0.199464       -0.361746      -0.068535       -0.464739  
4         -0.294350        0.028230      -0.226953       -0.475206  

[5 rows x 32 columns]

In [92]:
df_concat.shape

(199178, 32)